fontscan/substitutions.go

```diff
@@ -0,0 +1,254 @@
+func (fl familyList) insertAfter(element *list.Element, families []string) {
+	for _, s := range families {
+		element = fl.List.InsertAfter(s, element)
+	}
+}
+
+// insertBefore inserts families right before element
+func (fl familyList) insertBefore(element *list.Element, families []string) {
+	L := len(families)
+	for i := range families {
+		element = fl.List.InsertBefore(families[L-1-i], element)
+	}
+}
+
+func (fl familyList) replace(element *list.Element, families []string) {
+	fl.insertAfter(element, families)
+	fl.List.Remove(element)
+}
+
+// ----- substitutions ------
+
+// where to insert the families with respect to
+// the current list
+type substitutionOp uint8
+
+const (
+	opAppend substitutionOp = iota
+	opAppendLast
+	opPrepend
+	opPrependFirst
+	opReplace
+)
+
+type substitutionTest interface {
+	// returns a non nil element if the substitution should
+	// be applied
+	// for opAppendLast and opPrependFirst an arbitrary non nil element
+	// could be returned
+	test(list familyList) *list.Element
+
+	// return a copy where families have been normalize
+	// to their no blank no case version
+	normalize() substitutionTest
+}
+
+// a family in the list must equal 'mf'
+type familyEquals string
+
+func (mf familyEquals) test(list familyList) *list.Element {
+	return list.elementEquals(string(mf))
+}
+
+func (mf familyEquals) normalize() substitutionTest {
+	return familyEquals(meta.NormalizeFamily(string(mf)))
+}
+
+// a family in the list must contain 'mf'
+type familyContains string
+
+func (mf familyContains) test(list familyList) *list.Element {
+	return list.elementContains(string(mf))
+}
+
+func (mf familyContains) normalize() substitutionTest {
+	return familyContains(meta.NormalizeFamily(string(mf)))
+}
+
+// the family list has no "serif", "sans-serif" or "monospace" generic fallback
+type noGenericFamily struct{}
+
+func (noGenericFamily) test(list familyList) *list.Element {
+	for l := list.List.Front(); l != nil; l = l.Next() {
+		switch l.Value.(string) {
+		case "serif", "sans-serif", "monospace":
+			return nil
+		}
+	}
+	return list.List.Front()
+}
+
+func (noGenericFamily) normalize() substitutionTest {
+	return noGenericFamily{}
+}
+
+// one family must equals `family`, and the queried language
+// must equals `lang`
+type langAndFamilyEqual struct {
+	lang   language.Language
+	family string
+}
+
+// TODO: for now, these tests language base tests are ignored
+func (langAndFamilyEqual) test(list familyList) *list.Element {
+	return nil
+}
+
+func (t langAndFamilyEqual) normalize() substitutionTest {
+	t.family = meta.NormalizeFamily(t.family)
+	return t
+}
+
+// one family must equals `family`, and the queried language
+// must contains `lang`
+type langContainsAndFamilyEquals struct {
+	lang   language.Language
+	family string
+}
+
+// TODO: for now, these tests language base tests are ignored
+func (langContainsAndFamilyEquals) test(list familyList) *list.Element {
+	return nil
+}
```

Comment:
Similarly, leaving this unimplemented prevents us from applying a substitution that has a set of possible languages and a family requirement?